### PR TITLE
chore: Add SessionState to MockContextProvider just like SessionContextProvider

### LINKED ISSUE #11531: chore: Add `SessionState` to `MockContextProvider` just like `SessionContextProvider`
### Is your feature request related to a problem or challenge?

              I prefer not to re-introduce `with_expr_planner` back for context provider.

It is better to use session state in context, and add planner into session state.

Similar to SessionContextProvider
```rust
struct SessionContextProvider<'a> {
    state: &'a SessionState,
    ...
}
```

_Originally posted by @jayzhan211 in https://github.com/apache/datafusion/pull/11487#discussion_r1682349054_
            
This is cleanup task to remove functions like `with_udaf` `with_expr_planner` from `MockContextProvider` and make it consistent with public interfaces. We need to update all the testcases which are affected by this change.

### Describe the solution you'd like

_No response_

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by dharanad**
## Which issue does this PR close?

Closes #11531

### CODE REVIEW

**Comment by dharanad** [comment]
Cleaning up some backlog
cc: @jayzhan211

**Review by jayzhan211 — APPROVED** [review]
👍

File: datafusion/sql/tests/common/mod.rs [hunk 1]
```diff
@@ -50,36 +50,40 @@ impl Display for MockCsvType {
...
+impl MockSessionState {
+    pub fn with_expr_planner(mut self, expr_planner: Arc<dyn ExprPlanner>) -> Self {
+        self.expr_planners.push(expr_planner);
+        self
     }
 
-    #[allow(dead_code)]
-    pub(crate) fn with_udf(mut self, udf: ScalarUDF) -> Self {
-        self.udfs.insert(udf.name().to_string(), Arc::new(udf));
+    pub fn with_scalar_functions(mut self, scalar_functions: Arc<ScalarUDF>) -> Self {
```
Comment (jayzhan211) [inline datafusion/sql/tests/common/mod.rs:66]:
```suggestion
    pub fn with_scalar_function(mut self, scalar_function: Arc<ScalarUDF>) -> Self {
```

File: datafusion/sql/tests/common/mod.rs [hunk 2]
```diff
@@ -50,36 +50,40 @@ impl Display for MockCsvType {
...
+        self.scalar_functions
+            .insert(scalar_functions.name().to_string(), scalar_functions);
         self
     }
 
-    pub(crate) fn with_udaf(mut self, udaf: Arc<AggregateUDF>) -> Self {
+    pub fn with_aggregate_functions(
```
Comment (jayzhan211) [inline datafusion/sql/tests/common/mod.rs:72]:
```suggestion
    pub fn with_aggregate_function(
```

File: datafusion/sql/tests/common/mod.rs (hunk 1 above)
Comment (dharanad) [reply datafusion/sql/tests/common/mod.rs:66]: I will refactor

File: datafusion/sql/tests/sql_integration.rs [hunk 3]
```diff
@@ -2739,39 +2742,44 @@ fn logical_plan_with_dialect_and_options(
     dialect: &dyn Dialect,
     options: ParserOptions,
 ) -> Result<LogicalPlan> {
-    let context = MockContextProvider::default()
-        .with_udf(unicode::character_length().as_ref().clone())
-        .with_udf(string::concat().as_ref().clone())
-        .with_udf(make_udf(
+    let state = MockSessionState::default()
+        .with_scalar_function(Arc::new(unicode::character_length().as_ref().clone()))
```
Comment (alamb) [inline datafusion/sql/tests/sql_integration.rs:2746]: ❤️

**Review by alamb — APPROVED** [review]
I think this looks really nice to me -- thank you @dharanad

**Pull request merged**